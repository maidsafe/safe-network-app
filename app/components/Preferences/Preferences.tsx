import React, { Component } from 'react';
import List from '@material-ui/core/List';

import { PreferenceItem } from './PreferenceItem';

import { UserPreferences } from '$Definitions/application.d';
import { generateRandomString } from '$Utils/app_utils';
import {
    launchOnLogin,
    pinLaunchpadToMenu,
    releaseLaunchpadFromMenu,
    storeUserPreferences
} from '$App/actions/launchpad_actions';

interface Props {
    userPreferences: UserPreferences;
    onChange: Function;
    requiredItems?: { [item: string]: boolean };
}

export class Preferences extends Component<Props> {
    static defaultProps = {
        userPreference: {},
        requiredItems: {
            autoUpdate: true,
            pinToMenuBar: true,
            launchOnStart: true,
            showDeveloperApps: true,
            warnOnAccessingClearnet: true
        },
        onChange: () => {}
    };

    public static changeCompleted( userPreferences ) {
        // Enable or disable auto launch
        launchOnLogin( userPreferences.launchOnStart );

        // switch between standard or tray window
        if ( userPreferences.pinToMenuBar ) {
            pinLaunchpadToMenu();
        } else {
            releaseLaunchpadFromMenu();
        }

        // Save user preference
        storeUserPreferences( userPreferences );
    }

    private handleChange = ( name: string, changedStatus: boolean ) => {
        const { userPreferences, onChange } = this.props;
        switch ( name ) {
            case 'pinToMenuBar':
                if ( changedStatus ) {
                    pinLaunchpadToMenu();
                } else {
                    releaseLaunchpadFromMenu();
                }
                break;
            case 'launchOnStart':
                launchOnLogin( changedStatus );
                break;
            default:
                break;
        }

        userPreferences[name] = changedStatus;
        if ( typeof onChange === 'function' ) {
            onChange( userPreferences );
        }
    };

    render() {
        const { userPreferences, requiredItems } = this.props;
        const requiredItemArray = Object.keys( requiredItems );

        return (
            <List aria-label="Preferences">
                {Object.keys( userPreferences ).map( ( userPreference ) => {
                    if ( !requiredItemArray.includes( userPreference ) )
                        return null;

                    return (
                        <PreferenceItem
                            key={generateRandomString()}
                            name={userPreference}
                            status={userPreferences[userPreference]}
                            onChange={this.handleChange}
                            disabled={!requiredItems[userPreference]}
                        />
                    );
                } )}
            </List>
        );
    }
}
