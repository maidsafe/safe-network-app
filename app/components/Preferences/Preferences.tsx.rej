diff a/app/components/Preferences/Preferences.tsx b/app/components/Preferences/Preferences.tsx	(rejected hunks)
@@ -67 +67 @@ export class Preferences extends Component<Props> {
-        if (typeof onChange === 'function') {
+        if ( typeof onChange === 'function' ) {
