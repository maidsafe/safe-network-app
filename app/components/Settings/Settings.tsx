import React, { Component } from 'react';
import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import ArrowBack from '@material-ui/icons/ArrowBack';

import { Preferences } from '$Components/Preferences';
import { UserPreferences } from '$Definitions/application.d';

interface Props {
    userPreferences: UserPreferences;
    setUserPreferences: Function;
}

// eslint-disable-next-line unicorn/prevent-abbreviations
export class Settings extends Component<Props> {
    componentWillUnmount() {
        Preferences.changeCompleted( this.props.userPreferences );
    }

    render() {
        const { userPreferences, setUserPreferences } = this.props;
        return (
            <Grid container>
                <Box>
                    <Toolbar>
                        <IconButton
                            edge="start"
                            color="inherit"
                            aria-label="Back"
                        >
                            <ArrowBack fontSize="inherit" />
                        </IconButton>
                        <Typography>Settings</Typography>
                    </Toolbar>
                </Box>
                <Grid item xs={12}>
                    <Preferences
                        userPreferences={userPreferences}
                        onChange={setUserPreferences}
                    />
                </Grid>
            </Grid>
        );
    }
}
