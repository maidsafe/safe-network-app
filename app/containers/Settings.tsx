import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Settings } from '$Components/Settings';

import { setUserPreferences } from '$Actions/launchpad_actions';
import {
    storeUserPreferences,
    autoLaunch,
    pinToTray
} from '$Actions/alias/launchpad_actions';

const mapStateToProperties = ( state ) => {
    return {
        userPreferences: state.launchpad.userPreferences
    };
};

const mapDispatchToProperties = ( dispatch ) => {
    console.log( 'storeUserPreferences', storeUserPreferences );
    const actions = {
        setUserPreferences,
        storeUserPreferences,
        autoLaunch,
        pinToTray
    };
    return bindActionCreators( actions, dispatch );
};

export const SettingsPage: React.ComponentClass = connect(
    mapStateToProperties,
    mapDispatchToProperties
)( Settings );
