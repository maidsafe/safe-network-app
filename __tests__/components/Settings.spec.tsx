import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import IconButton from '@material-ui/core/IconButton';
import Typography from '@material-ui/core/Typography';

import { Settings } from '$Components/Settings';
import { Preferences } from '$Components/Preferences';

const shallow = createShallow();

// eslint-disable-next-line unicorn/prevent-abbreviations
const shallowSetup = ( propOverrides? ) => {
    // eslint-disable-next-line unicorn/prevent-abbreviations
    const props = Object.assign(
        {
            userPreferences: {
                autoUpdate: false,
                pinToMenuBar: false,
                launchOnStart: false,
                showDeveloperApps: false,
                warnOnAccessingClearnet: false
            },
            setUserPreferences: jest.fn(),
            history: {
                push: jest.fn()
            }
        },
        propOverrides
    );

    const wrapper = shallow( <Settings {...props} /> );

    return {
        props,
        wrapper
    };
};

describe( 'Settings Page', () => {
    it( 'render', () => {
        const { wrapper } = shallowSetup();
        expect( wrapper ).toMatchSnapshot();
    } );

    it( 'should have title `Settings`', () => {
        const expectedTitle = 'Settings';
        const { wrapper } = shallowSetup();
        const titleElement = wrapper
            .find( Typography )
            .filterWhere( ( n ) => n.prop( 'aria-label' ) === 'title' );
        expect( titleElement.text() ).toEqual( expectedTitle );
    } );

    it( 'should load preferences once', () => {
        const { wrapper } = shallowSetup();
        expect( wrapper.find( Preferences ) ).toHaveLength( 1 );
    } );

    it( 'should go to home on clicking back button', () => {
        const { wrapper, props } = shallowSetup();
        wrapper.find( IconButton ).simulate( 'click' );
        expect( props.history.push ).toHaveBeenCalled();
        expect( props.history.push ).toHaveBeenCalledWith( '/' );
    } );
} );
