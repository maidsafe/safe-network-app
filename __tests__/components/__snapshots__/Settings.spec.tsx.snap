// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Page render 1`] = `ShallowWrapper {}`;
